lib/ansible/modules/cloud/vmware/vmware_guest_search.py

```diff
@@ -0,0 +1,147 @@
+  vmware_guest_search:
+    hostname: '{{ vcenter_server }}'
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    datacenter: '{{ vcenter_datacenter }}'
+    validate_certs: no
+    name: base-vm
+  delegate_to: localhost
+  register: facts
+'''
+
+RETURN = """
+vms:
+    description: A list of virtual machine names that start with the provided name
+    returned: always
+    type: list
+"""
+
+try:
+    import pyVmomi
+    from pyVmomi import vim
+except ImportError:
+    pass
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_text
+from ansible.module_utils.vmware import PyVmomi, vmware_argument_spec, find_datacenter_by_name
+
+
+class PyVmomiHelper(PyVmomi):
+    def __init__(self, module):
+        super(PyVmomiHelper, self).__init__(module)
+
+    def find_folder(self, root, searchpath):
+        """ Walk inventory objects one position of the searchpath at a time """
+        # split the searchpath so we can iterate through it
+        paths = [x.replace('/', '') for x in searchpath.split('/')]
+        paths_total = len(paths) - 1
+        position = 0
+
+        # recursive walk while looking for next element in searchpath
+
+        while root and position <= paths_total:
+            change = False
+            if hasattr(root, 'childEntity'):
+                for child in root.childEntity:
+                    if child.name == paths[position]:
+                        root = child
+                        position += 1
+                        change = True
+                        break
+            elif isinstance(root, vim.Datacenter):
+                if hasattr(root, 'vmFolder'):
+                    if root.vmFolder.name == paths[position]:
+                        root = root.vmFolder
+                        position += 1
+                        change = True
+            else:
+                root = None
+
+            if not change:
+                root = None
+
+        return root
+
+
+def main():
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(
+        name=dict(type='str', required=True),
+        datacenter=dict(type='str', required=True),
+    )
+    module = AnsibleModule(argument_spec=argument_spec)
+    try:
+        pyv = PyVmomiHelper(module)
+
+        dc = find_datacenter_by_name(pyv.content, module.params['datacenter'])
+        if dc is None:
+            module.fail_json(msg="Failed to find the datacenter %s" % module.params['datacenter'])
+        folder = dc.vmFolder
+
+        name = module.params['name']
+        if "/" in name:
+            paths = name.split("/")
+            paths.pop()
+            paths = "/".join(paths)
+            name = name.split("/").pop()
+            folder = pyv.find_folder(dc.vmFolder, paths)
+            if folder is None:
+                module.fail_json(msg="Unable to find folder: %s" % paths)
+        vms = []
+        for vm in folder.childEntity:
```

Comment:
This does not work if you have datacenter nested inside a folder. For example, datacenter `DC0` is nested under folder `F0` then this will always returns `F0`.

Reply:
Will have a look - do you have a failing test case I can re-use?